File: ccpl/2015/r8/A_wheels/Main.java
import java.io.*;
import java.math.*;
import java.util.*;

public class Main {

	public static void main(String... args) throws IOException {
		InputStreamReader isr = new InputStreamReader(System.in);
		BufferedReader in = new BufferedReader(isr);
		long T = Integer.parseInt(in.readLine());
		for (int i = 0; i < T; i++) {
			int n = Integer.parseInt(in.readLine());
			Wheel[] wheels = new Wheel[n];
			for (int j = 0; j < n; j++) {
				String[] vals = in.readLine().split(" ");
				wheels[j] = new Wheel(
					Integer.parseInt(vals[0]),
					Integer.parseInt(vals[1]),
					Integer.parseInt(vals[2])
				);
			}

			// Force in wheel #1
			wheels[0].move = 1;

			String ans = getAns(wheels);
			System.out.println(ans);
		}
	}

	static String getAns(Wheel[] wheels) {
		Stack<Integer> S = new Stack<Integer>();
		S.push(0);
		while (!S.empty()) {
			int u = S.pop();
			for (int i = 0; i < wheels.length; i++) {
				if (theyTouch(wheels[i], wheels[u]) && wheels[i].move == 0) {
					wheels[i].move = -wheels[u].move;
					S.add(i);
				}
			}
		}

		StringBuilder ans = new StringBuilder();
		for (Wheel w : wheels) {
			if (w.move == 0) {
				ans.append("not moving");
			} else {
				long gcd = BigInteger.valueOf(wheels[0].r).gcd(
					BigInteger.valueOf(w.r)).longValue();
				ans.append(wheels[0].r / gcd);
				ans.append((w.r > gcd) ? "/" + (w.r / gcd) : "");
				ans.append((w.move == 1) ? " clockwise" : " counterclockwise");
			}
			ans.append("\n");
		}
		return ans.substring(0, ans.length() - 1);
	}

	static boolean theyTouch(Wheel wi, Wheel wj) {
		long distX   = (wi.x - wj.x) * (wi.x - wj.x);
		long distY   = (wi.y - wj.y) * (wi.y - wj.y);
		long rSquare = (wi.r + wj.r) * (wi.r + wj.r);
		return (distX + distY) == rSquare;
	}

	static class Wheel {
		long x, y, r;
		int move;
		public Wheel(long x, long y, long r) {
			this.x = x;
			this.y = y;
			this.r = r;
		}
	}
}

File: ccpl/2015/r8/A_wheels/Main.c
#include <stdio.h>

void getAns(int x[], int y[], int r[], int move[], int n);
int theyTouch(int x[], int y[], int r[], int i, int j);
int getGcd(int a, int b);

int main() {
	int i, j, n, T, x[1000], y[1000], r[1000], move[1000];

	scanf("%d", &T);
	for (i = 0; i < T; i++) {
		/* Init arrays. */
		for (j = 0; j < 1000; j++) {
			x[j] = 0; y[j] = 0; r[j] = 0; move[j] = 0;
		}

		scanf("%d", &n);
		for (j = 0; j < n; j++) {
			int xAux, yAux, rAux;
			scanf("%d %d %d", &xAux, &yAux, &rAux);
			x[j] = xAux; y[j] = yAux; r[j] = rAux;
		}

		/* Initial force. */
		move[0] = 1;

		getAns(x, y, r, move, n);
	}

	return 0;
}

void getAns(int x[], int y[], int r[], int move[], int n) {
	int i, l, stack[n];
	stack[0] = 0; l = 1;
	while (l > 0) {
		int u = stack[l-1]; l--;
		for (i = 0; i < n; i++) {
			if (theyTouch(x, y, r, u, i) == 1 && move[i] == 0) {
				int old = move[i];
				move[i] = -move[u];
				stack[l] = i; l++;
			}
		}
	}

	for (i = 0; i < n; i++) {
		if (move[i] == 0) {
			printf("not moving\n");
		} else {
			int gcd = getGcd(r[0], r[i]);
			if (r[i] == gcd) {
				if (move[i] == 1) {
					printf("%d clockwise\n", r[0] / gcd);
				} else {
					printf("%d counterclockwise\n", r[0] / gcd);
				}
			} else {
				if (move[i] == 1) {
					printf("%d/%d clockwise\n", r[0] / gcd, r[i] / gcd);
				} else {
					printf("%d/%d counterclockwise\n", r[0] / gcd, r[i] / gcd);
				}
			}
		}
	}
}

int theyTouch(int x[], int y[], int r[], int i, int j) {
	int distX = (x[i] - x[j]) * (x[i] - x[j]);
	int distY = (y[i] - y[j]) * (y[i] - y[j]);
	int rSquare = (r[i] + r[j]) * (r[i] + r[j]);
	if (distX + distY == rSquare) {
		return 1;
	} else {
		return 0;
	}
}

int getGcd(int a, int b) {
	if (a == 0) {
		return b;
	} else {
		return getGcd(b % a, a);
	}
}
